allennlp/evaluation/evaluator.py

```diff
@@ -0,0 +1,231 @@
+        batch_weight_key : `str`, optional (default=`None`)
+            If given, this is a key in the output dictionary for each batch that specifies how to weight
+            the loss for that batch.  If this is not given, we use a weight of 1 for every batch.
+        metrics_output_file : `Union[str, PathLike]`, optional (default=`None`)
+            Optional path to write the final metrics to.
+        predictions_output_file : `Union[str, PathLike]`, optional (default=`None`)
+            Optional path to write the predictions to.
+
+        # Returns
+
+        metrics: `Dict[str, Any]`
+            The metrics from evaluating the file.
+        """
+        check_for_gpu(self.cuda_device)
+        data_loader.set_target_device(int_to_device(self.cuda_device))
+        output_file = Path(output_file) if output_file is not None else None
+        if predictions_output_file is not None:
+            predictions_file = Path(predictions_output_file).open("w", encoding="utf-8")
+        else:
+            predictions_file = None  # type: ignore
+
+        model_postprocess_function = getattr(model, "make_output_human_readable", None)
+
+        with torch.no_grad():
+            model.eval()
+
+            iterator = iter(data_loader)
+            logger.info("Iterating over dataset")
+            generator_tqdm = Tqdm.tqdm(iterator)
+            # Number of batches in instances.
+            batch_count = 0
+            # Number of batches where the model produces a loss.
+            loss_count = 0
+            # Cumulative weighted loss
+            total_loss = 0.0
+            # Cumulative weight across all batches.
+            total_weight = 0.0
+
+            for batch in generator_tqdm:
+                batch_count += 1
+                batch = nn_util.move_to_device(batch, self.cuda_device)
+                output_dict = model(**batch)
+                loss = output_dict.get("loss")
+
+                metrics = model.get_metrics()
+
+                if loss is not None:
+                    loss_count += 1
+                    if batch_weight_key:
+                        weight = output_dict[batch_weight_key].item()
+                    else:
+                        weight = 1.0
+
+                    total_weight += weight
+                    total_loss += loss.item() * weight
+                    # Report the average loss so far.
+                    metrics["loss"] = total_loss / total_weight
+
+                description = (
+                    ", ".join(
+                        [
+                            "%s: %.2f" % (name, value)
+                            for name, value in metrics.items()
+                            if not name.startswith("_")
+                        ]
+                    )
+                    + " ||"
+                )
+                generator_tqdm.set_description(description, refresh=False)
+
+                # TODO(gabeorlanski): Add in postprocessing the batch for token
```

Comment:
Would this be something like calculating BLEU scores over the predicted tokens?

Reply:
Yea that would be for BLEU or metrics along those lines. It might be easier to just decode before, but that would require rewriting many of the already defined metrics to accept tokens rather than tensors